File: demo/scala/Clock.scala
//
//  Clock.scala - the Scala demo client, riding the Java SDK.
//
//  Copyright 2026 Saxon Herschel Nicholls
//  SPDX-License-Identifier: MIT
//
//  Scala needs no SDK of its own for the same reason Kotlin does not: it
//  IS the JVM, and the Java client - the builder, the java.util.logging
//  handler, the uncaught-exception hook - is idiomatic enough to call
//  directly. What this file demonstrates is exactly that: zero wrapper,
//  zero glue, one import.
//
//    javac -d /tmp/sls $(find sdk/java -name '*.java')
//    scalac -classpath /tmp/sls -d /tmp/sls demo/scala/Clock.scala
//    sh demo/scala/run.sh /tmp/sls            # java + the Scala runtime jars
//    (Scala >= 3.5 replaced the classic runner with scala-cli, which no
//    longer launches a bare class - run.sh picks whichever way works)
//
//  The same clock every other demo client runs, once a second on
//  scala.clock: tick at INFO, a DEBUG pricing pass, a real ERROR when the
//  pricer meets the symbol it has no rate for, an uptime metric.
//

import com.snicholls.superlog.SuperLog

import java.time.ZoneOffset
import java.time.format.DateTimeFormatter
import java.time.Instant

object Clock {

  private val Hms =
    DateTimeFormatter.ofPattern("HH:mm:ss'Z'").withZone(ZoneOffset.UTC)

  private val Rates = Map("BTC" -> 64000.0, "ETH" -> 3200.0)

  def main(args: Array[String]): Unit = {
    val maxTicks = args.sliding(2).collectFirst {
      case Array("--ticks", n) => n.toInt
    }.getOrElse(0)

    val log = SuperLog.builder()
      .topic("scala.clock")
      .app("clock")
      .url(sys.env.getOrElse("SUPER_LOG_URL", "http://127.0.0.1:7333"))
      .development(true)
      .production(false)
      .build()
    log.installUncaughtHandler()

    println(s"superlog: scala clock -> ${log.status().topic()}")
    log.info("scala clock up - one line a second",
             SuperLog.fields("scala", util.Properties.versionNumberString))

    var tick = 0
    while (maxTicks == 0 || tick < maxTicks) {
      tick += 1
      log.info(s"tick $tick - the time is ${Hms.format(Instant.now())}",
               SuperLog.fields("tick", Int.box(tick)))

      // Honestly wrong every 7th tick, the same staged failure as every
      // other clock, so one error lines up across every language.
      val symbol = if (tick % 7 == 0) "DOGE" else "BTC"
      Rates.get(symbol) match {
        case Some(rate) =>
          log.debug(s"pricing pass $tick",
                    SuperLog.fields("symbol", symbol, "price", Double.box(rate * 2)))
        case None =>
          log.error(s"pricing failed on tick $tick: no rate for $symbol",
                    SuperLog.fields("symbol", symbol, "tick", Int.box(tick)))
      }

      if (tick % 5 == 0)
        log.debug("clock.uptime_s", SuperLog.fields("value", Int.box(tick)))

      Thread.sleep(1000)
    }
    log.info("scala clock stopping")
  }
}

File: demo/haskell/clock.hs
--
--  clock.hs - the Haskell demo client.
--
--  Copyright 2026 Saxon Herschel Nicholls
--  SPDX-License-Identifier: MIT
--
--  The same clock every other demo client runs, once a second on
--  haskell.clock: the tick at INFO, a DEBUG pricing pass, a real ERROR
--  when the pricer meets the symbol it has no rate for, and an uptime
--  metric every fifth tick.
--
--    ghc -DDEVELOPMENT -isdk/haskell -o clock demo/haskell/clock.hs && ./clock
--    ./clock --ticks 6        # stop after six, for scripts
--

import Control.Monad (forM_, when)
import Control.Concurrent (threadDelay)
import Data.Time.Clock (getCurrentTime)
import Data.Time.Format (defaultTimeLocale, formatTime)
import System.Environment (getArgs)

import SuperLog

rates :: [(String, Double)]
rates = [("BTC", 64000), ("ETH", 3200)]

main :: IO ()
main = do
  args <- getArgs
  let maxTicks = case dropWhile (/= "--ticks") args of
        (_ : n : _) -> maybe 0 id (readMaybe n)
        _ -> 0
      readMaybe s = case reads s of { [(v, "")] -> Just v; _ -> Nothing }

  lg <- newLog "haskell.clock" "clock"
  putStrLn "superlog: haskell clock -> haskell.clock"
  info lg "haskell clock up - one line a second" [("compiler", "ghc")]

  let ticks = if maxTicks > 0 then [1 .. maxTicks] else [1 ..]
  forM_ ticks $ \tick -> do
    now <- getCurrentTime
    let hms = formatTime defaultTimeLocale "%H:%M:%SZ" now
    info lg ("tick " ++ show tick ++ " - the time is " ++ hms)
            [("tick", show tick)]

    -- Honestly wrong every 7th tick, the same staged failure as every
    -- other clock, so one error lines up across every language.
    let symbol = if tick `mod` 7 == 0 then "DOGE" else "BTC"
    case lookup symbol rates of
      Just r -> debug lg ("pricing pass " ++ show tick)
                         [("symbol", symbol), ("price", show (r * 2))]
      Nothing -> err lg ("pricing failed on tick " ++ show tick ++
                         ": no rate for " ++ symbol)
                        [("symbol", symbol), ("tick", show tick)]

    when (tick `mod` 5 == 0) $ metric lg "clock.uptime_s" (fromIntegral tick)
    flushLog lg
    threadDelay 1000000

  info lg "haskell clock stopping" []
  flushLog lg
